Mathlib/Tactic/DeriveFintype.lean

```diff
@@ -0,0 +1,315 @@
+    let mut toFun ← `(term| fun $toFunAlts:matchAlt*)
+    let mut invFun ← `(term| fun $invFunAlts:matchAlt*)
+    if indVal.numCtors == 0 then
+      -- Empty matches don't elaborate, so use `nomatch` here.
+      toFun ← `(term| fun x => nomatch x)
+      invFun ← `(term| fun x => nomatch x)
+    let equivBody ← `(term|
+      { toFun := $toFun,
+        invFun := $invFun,
+        right_inv := by intro x; cases x <;> rfl
+        left_inv := by intro x; $pf:tactic })
+
+    return { proxyName := proxyName
+             equiv := equivBody }
+
+/--
+The term elaborator `derive_fintype% α` tries to synthesize a `Fintype α` instance
+using all the assumptions in the local context. It only works when `α` is an inductive
+type that the `Fintype` deriving handler can normally process. The handler makes use of the
+expected type, so `(derive_fintype% _ : Fintype α)` works as well.
+
+A side effect of this is that it defines a type named `fintypeProxy` in the namespace associated
+to `α`, just like the `Fintype` deriving handler normally would for non-enum types.
+-/
+elab "derive_fintype% " t:term : term <= expectedType => do
+  let type ← Term.elabType t
+  let f ← Term.elabType (← `(Fintype $(← Term.exprToSyntax type)))
+  unless ← isDefEq expectedType f do
+    throwError "Could not unify expected type{indentExpr expectedType}\nwith{indentExpr f}"
+  let mut type ← instantiateMVars type
+  if type.hasExprMVar then
+    Term.synthesizeSyntheticMVars
+    type ← instantiateMVars type
+    if type.hasExprMVar then
+      throwError "Provided type {type} has metavariables"
+  type ← whnf type
+  let .const declName _ := type.getAppFn
+    | throwError "{type} is not a constant or constant application"
+  let indVal ← getConstInfoInduct declName
+  let data ← mkProxyEquiv indVal
+  let proxy ← mkAppOptM data.proxyName (type.getAppArgs.map .some)
+  let equivType ← mkAppM ``Equiv #[proxy, type]
+  let equiv ← Term.elabTerm data.equiv equivType
+  return ← mkAppM ``Fintype.ofEquiv #[proxy, equiv]
+
+/-
+Creates a `Fintype` instance by adding additional `Fintype` and `Decidable` instance arguments
+for every type and prop parameter of the type, then use the `derive_fintype%` elaborator.
+-/
+def mkFintype (declName : Name) : CommandElabM Bool := do
+  let indVal ← getConstInfoInduct declName
+  let cmd ← liftTermElabM do
+    let header ← Deriving.mkHeader `Fintype 0 indVal
+    let binders' ← Deriving.mkInstImplicitBinders `Decidable indVal header.argNames
+    let instCmd ← `(command|
+      instance $header.binders:bracketedBinder* $(binders'.map TSyntax.mk):bracketedBinder* :
+          Fintype $header.targetType := derive_fintype% _)
+    return instCmd
+  trace[Elab.Deriving.fintype] "instance command:\n{cmd}"
+  elabCommand cmd
+  return true
+
+/-- Derive a `Fintype` instance for enum types. These come with a `toCtorIdx` function.
+The strategy is to (1) create a list `enumList` of all the constructors, (2) prove that this
```

Comment:
Perhaps it would make sense to implement a `FinEnum` instance derive handler here too?

Reply:
I'd suggest making a separate deriving handler for this, since the `List` isn't so useful for the equivalence.

Note that there's already an inverse function for `toCtrIdx` defined by the `DecidableEq` deriving handler for enum types. https://github.com/leanprover/lean4/blob/master/src/Lean/Elab/Deriving/DecEq.lean#L110

Reply:
Is the list useful for anything? Why do we want a special path for enum types?

Reply:
It's significantly less expensive in both space and time, if we're going to do any computations with it. 63 lines seems like a small price to pay for that.

Reply:
When is the cost paid for computing the elements of a fintype? Every time you use `finset.sum`, or just once when the instance is added to the environment?

Reply:
It depends on what kind of computing you're doing. I think if it's in the compiled code, an instance with no arguments (like for types with no parameters) will be evaluated at program initialization. But otherwise, for VM evaluation or symbolic evaluation (like for Decidable-like inference or `fin_cases`) I think it's more like the elements get computed from scratch at every use.